source/simulator/core.cc

```diff
@@ -1678,19 +1678,38 @@ namespace aspect
             free_surface->execute ();
 
           assemble_advection_system (AdvectionField::temperature());
-          build_advection_preconditioner(AdvectionField::temperature(),
-                                         T_preconditioner);
-          solve_advection(AdvectionField::temperature());
+          if (system_rhs.block(AdvectionField::temperature().block_index(introspection)).l2_norm() <= std::numeric_limits<double>::min())
```

Comment:
I am not sure it is needed, but with your suggestion I get
```
a.cc: In function ‘int main()’:
a.cc:14:26: warning: comparing floating point with == or != is unsafe [-Wfloat-equal]
   std::cout << "" << (a==0.0);
```
with g++ 6.1. Any suggestions?